### ISSUE TITLE
pipeline-core is uninstallable on Linux without a C compiler: numpy<2 has no cp313 wheel

### ISSUE THREAD

**Issue by ymohanty** [Data-Policy-and-Innovation-Centre/janasunani#134]
Found tonight when the redaction backfill could not install its environment on the CPU box.

## What happens

`uv sync --extra pipeline-core` fails on the box with:

```
numpy==1.26.4 ... Unknown compiler(s): [cc, gcc, clang, ...]
```

## Why

- `pyproject.toml:52` pins `numpy<2` on **pipeline-core**, resolving to `numpy==1.26.4`
- The repo pins `.python-version = 3.13`
- `numpy==1.26.4` predates Python 3.13 and publishes **no cp313 wheel**. `uv.lock` lists **zero** wheels for it — sdist only

So on every platform this builds from source. It succeeds on macOS because Xcode ships clang, and fails on the box because it has no compiler. The pin's own comment says it follows the DSI repo (`numpy<2, opencv<4.12, transformers>=4.57,<5`) — it is inherited, not derived from what these stages need.

**This is not box-specific.** Any Linux machine without build tooling cannot install this extra: the GPU box in week 2, a fresh CI runner, a new contributor on Linux.

## The PII path does not need the pin

Evidence from this machine, running the merged `#69` code:

- local venv has **numpy 2.5.0**
- `presidio_analyzer`, `spacy` and `en_core_web_sm` all import
- `tests/test_redact_grievance.py` — **18 passed, 0 skipped**, including `test_real_redactor_removes_a_mobile_number`, which exercises the real analyzer rather than the stub

So redaction runs correctly on numpy 2.x.

## What the pin is probably protecting

Not redaction. The likely reasons are the other pipeline-core stages:

- the pickled estimators (`page_split` pkl, the label encoder), pinned alongside `scikit-learn>=1.8,<1.9` to "match the era of the pickled estimators" — unpickling across a numpy major is the classic breakage
- `opencv-python<4.12`
- `xgboost` / `scikit-image` expectations

Whoever fixes this should establish which of those actually constrain numpy, rather than assuming the whole extra needs it.

## Options

1. **Split the extra.** A `pii` extra with just presidio, spacy and the model — no numpy ceiling, all wheels available. Cleanest, and matches the existing design where stages import lazily and an environment installs only what it runs.
2. **Relax `numpy<2`** across pipeline-core, after verifying the pickled estimators still load and the ViT/page-split stages still behave. Bigger blast radius, needs real testing.
3. **Provision build tooling** on every Linux host. Treats the symptom, is slow, and `numpy==1.26.4` may not even compile against the 3.13 C API.

**Option 1 is the demo-week answer**; option 2 is the right end state.

## Related

- #69 — the redaction job that hit this
- #71 — the backfill it blocks

**Comment by ymohanty** [comment]
Why isnt this on a docker container that already has gcc then? that should be what goes on the box (any box) no?

**Comment by ymohanty** [comment]
## Option 4: run the pipeline in a container, which is probably the right end state

Raised by the maintainer while this was blocking tonight's backfill, and it is a better framing than the three options above.

Every option in the issue body treats this as a dependency-resolution problem. It is really an **environment-reproducibility** problem, and the reason it bit is that the box builds its own environment from `uv.lock` and therefore needs whatever toolchain the lock implies. A container inverts that: the environment is built once, somewhere that has a compiler, and the box runs it.

### Why this fits here specifically

- **The repo already ships images.** `deploy/` builds and runs the api image, and #48 / #62 already dealt with resolving torch from the CPU index for it. This is an extension of an existing pattern, not new infrastructure.
- **It makes `analyzer_version` honest.** Tonight's rows are stamped `presidio-analyzer=2.2.364 spacy=3.8.14 en-core-web-sm=3.8.0` — but the box is on `numpy 2.4.6` against a lock that says `1.26.4`, because the extra was installed piecemeal to get around this bug. A pinned image removes that divergence class entirely, which matters when the output feeds a published scorecard (#67).
- **It fixes the same problem for the GPU box** in week 2 rather than solving it once per host.

### The design constraint that shapes it

The extras are **mutually exclusive by declaration** (`[tool.uv].conflicts`) because DeepSeek OCR needs an older `transformers` than the rest. So this is not one image: it is one image per extra, or at minimum a `pipeline-core` image separate from an `ocr-deepseek` one. That is a feature rather than a cost — it is the same reason stages import lazily today.

For the PII path a **slim image** would do: presidio, spacy and the model, no torch, no opencv, no xgboost. Small, fast to pull, and it is exactly the split option 1 proposes, just delivered as an artifact instead of an extra.

### The catch, and it is a real one

The image has to be built somewhere with a toolchain — normally CI. **GitHub Actions has produced no runs since 2026-08-06T19:47Z**, so today that would mean building locally and pushing to GHCR by hand. Not a blocker for the idea, but it is why this is not the fix for tonight.

### Where that leaves the options

- **Tonight:** already unblocked by installing the PII dependencies directly (done, run in progress)
- **Demo week:** option 1, split a `pii` extra, so the lock alone is installable on a plain Linux host
- **End state:** this — containerise the pipeline environments, one per conflicting extra, and stop asking hosts to build anything

Options 1 and 4 are complementary rather than alternatives: a narrower extra is what makes a slim image possible.
